### PR TITLE
Add retry of openQA jobs based on test variables

### PR DESCRIPTION
**Pull request by okurz**
Motivation: Similar as in
https://docs.microsoft.com/en-us/azure/devops/release-notes/2021/sprint-195-update#automatic-retries-for-a-task
and https://docs.gitlab.com/ee/ci/yaml/#retry and
https://github.com/marketplace/actions/retry-step

Acceptance criteria:
* **AC1:** openQA jobs are automatically retriggered with a maximum amount
  specified in test variables
* **AC2:** By default openQA tests apply no automatic retry

Acceptance tests:
* **AT1-1:** *Given* an unfinished openQA test with the special `RETRY`
  variable given with value `N` *When* the job fails *And* the current job is
  not already the N-th clone of an original job *Then* an automatic clone is
  created
* **AT1-2:** *Given* an unfinished openQA test with the special `RETRY`
  variable given with value `N` *When* the job fails *And* the current job is
  the N-th clone of an original job *Then* no automatic clone is created
* **AT2-1:** *Given* an unfinished openQA test with no special retry variable
  given *When* the job fails *Then* no automatic clone is created at this
  point (job done hooks or external events may still trigger clones as well
  as manual triggers)

Review hint: I suggest to review the commits individually. As desired I
can also split the pull request into multiple for easier review.

Related progress issue: https://progress.opensuse.org/issues/104007

### CODE REVIEW

**Comment by coolo** [comment]
Only looked at the documentation of it, but I think it's a reasonable extension to the ecosystem

**Review by perlpunk — APPROVED** [review]

**Review by kraih — APPROVED** [review]

File: docs/WritingTests.asciidoc [hunk 1]
```diff
@@ -299,6 +299,32 @@ WORKER_CLASS = desktop
 # WORKER_CLASS is not set
 --------------------------------------------------------------------------------
 
+=== Automatic retries of jobs
+
+You might encounter flaky openQA tests that fail sporadically. While the best
```
Comment (Martchus) [inline docs/WritingTests.asciidoc:304]: The ""While" implies there's a 2nd part within the same sentence but there isn't.

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 2]
```diff
@@ -2004,6 +2004,20 @@ sub packages_diff {
...
+    return $o ? 1 + $o->ancestors : 0;
+}
+
+sub handle_retry ($self) {
+    return undef unless my $retry = $self->settings_hash->{RETRY};
+    # strip any optional descriptions after a colon
+    $retry =~ s/:.*//;
+    my $ancestors = $self->ancestors;
+    log_debug('handle_retry: Found retry job ' . $self->id . ", try '$ancestors' of up to '$retry' retries");
+    return $ancestors < $retry;
```
Comment (Martchus) [inline lib/OpenQA/Schema/Result/Jobs.pm:2018]: This lacks input validation, e.g. `$retry` might not be a number here.

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 3]
```diff
@@ -2004,6 +2004,20 @@ sub packages_diff {
...
+    my $o = $self->origin;
+    return $o ? 1 + $o->ancestors : 0;
+}
+
+sub handle_retry ($self) {
+    return undef unless my $retry = $self->settings_hash->{RETRY};
+    # strip any optional descriptions after a colon
+    $retry =~ s/:.*//;
+    my $ancestors = $self->ancestors;
+    log_debug('handle_retry: Found retry job ' . $self->id . ", try '$ancestors' of up to '$retry' retries");
```
Comment (Martchus) [inline lib/OpenQA/Schema/Result/Jobs.pm:2017]: It would be more useful to add this as a comment (visible in the web UI).

**Comment by Martchus** [comment]
This PR should fix all mentioned issues: https://github.com/os-autoinst/openQA/pull/4422#pullrequestreview-922883535=

**Pull request merged**